<a href="https://colab.research.google.com/github/aryasovaalexandra/praktika/blob/main/%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('/content/crimes_against_women_2001-2014.csv')
df

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
0               0  ANDHRA PRADESH           ADILABAD  2001    50   
1               1  ANDHRA PRADESH          ANANTAPUR  2001    23   
2               2  ANDHRA PRADESH           CHITTOOR  2001    27   
3               3  ANDHRA PRADESH           CUDDAPAH  2001    20   
4               4  ANDHRA PRADESH      EAST GODAVARI  2001    23   
...           ...             ...                ...   ...   ...   
10672         832     Lakshadweep        Lakshadweep  2014     1   
10673         833     Lakshadweep  Total District(s)  2014     1   
10674         834      Puducherry           Karaikal  2014     3   
10675         835      Puducherry         Puducherry  2014     7   
10676         836      Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
0                            30            16   
1                            30             7   
2                            34            14   
3                            20            17   
4                            26            12   
...                         ...           ...   
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
0                                                    149     
1                                                    118     
2                                                    112     
3                                                    126     
4                                                    109     
...                                                  ...     
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                              34                                  175   
1                              24                                  154   
2                              83                                  186   
3                              38                                   57   
4                              58                                  247   
...                           ...                                  ...   
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  
...                     ...  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0  

[10677 rows x 11 columns]

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [ ]:
df.columns


Index(['Unnamed: 0', 'STATE/UT', 'DISTRICT', 'Year', 'Rape',
       'Kidnapping and Abduction', 'Dowry Deaths',
       'Assault on women with intent to outrage her modesty',
       'Insult to modesty of Women', 'Cruelty by Husband or his Relatives',
       'Importation of Girls'],
      dtype='object')

In [ ]:
df.head(5)

Unnamed: 0        STATE/UT       DISTRICT  Year  Rape  \
0           0  ANDHRA PRADESH       ADILABAD  2001    50   
1           1  ANDHRA PRADESH      ANANTAPUR  2001    23   
2           2  ANDHRA PRADESH       CHITTOOR  2001    27   
3           3  ANDHRA PRADESH       CUDDAPAH  2001    20   
4           4  ANDHRA PRADESH  EAST GODAVARI  2001    23   

   Kidnapping and Abduction  Dowry Deaths  \
0                        30            16   
1                        30             7   
2                        34            14   
3                        20            17   
4                        26            12   

   Assault on women with intent to outrage her modesty  \
0                                                149     
1                                                118     
2                                                112     
3                                                126     
4                                                109     

   Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                          34                                  175   
1                          24                                  154   
2                          83                                  186   
3                          38                                   57   
4                          58                                  247   

   Importation of Girls  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0

In [ ]:
df.tail(5)

Unnamed: 0     STATE/UT           DISTRICT  Year  Rape  \
10672         832  Lakshadweep        Lakshadweep  2014     1   
10673         833  Lakshadweep  Total District(s)  2014     1   
10674         834   Puducherry           Karaikal  2014     3   
10675         835   Puducherry         Puducherry  2014     7   
10676         836   Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0

In [ ]:
df.describe()

Unnamed: 0          Year          Rape  Kidnapping and Abduction  \
count  10677.000000  10677.000000  10677.000000              10677.000000   
mean    3871.570198   2007.697949     57.989885                 69.888358   
std     2816.369918      4.046874    214.230398                311.623450   
min        0.000000   2001.000000      0.000000                  0.000000   
25%     1009.000000   2004.000000      8.000000                  6.000000   
50%     3678.000000   2008.000000     22.000000                 20.000000   
75%     6347.000000   2011.000000     44.000000                 49.000000   
max     9016.000000   2014.000000   5076.000000              10626.000000   

       Dowry Deaths  Assault on women with intent to outrage her modesty  \
count  10677.000000                                       10677.000000     
mean      20.181699                                         113.539196     
std       98.276531                                         458.903951     
min        0.000000                                           0.000000     
25%        1.000000                                          10.000000     
50%        5.000000                                          34.000000     
75%       16.000000                                          85.000000     
max     2469.000000                                       10001.000000     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
count                10677.000000                         10677.000000   
mean                    27.419313                           209.224314   
std                    167.806797                           905.664362   
min                      0.000000                             0.000000   
25%                      0.000000                            11.000000   
50%                      2.000000                            50.000000   
75%                     12.000000                           144.000000   
max                   4970.000000                         23278.000000   

       Importation of Girls  
count          10677.000000  
mean               0.175330  
std                2.228637  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max               83.000000

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





1.   **mean**	3871.570198	2007.697949	57.989885	69.888358	20.181699	113.539196	27.419313	209.224314	0.175330
2.   **min**	0.000000	2001.000000	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000
**max**	9016.000000	2014.000000	5076.000000	10626.000000	2469.000000	10001.000000	4970.000000	23278.000000	83.000000



##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [ ]:
df['Insult to modesty of Women'] = df['Insult to modesty of Women'].astype(float)

df.DISTRICT = df.DISTRICT.astype("category")

df.Rape= df.Rape.astype("int")
df.info()
#Вывести типы колонок тут



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10677 entries, 0 to 10676
Data columns (total 11 columns):
 #   Column                                               Non-Null Count  Dtype   
---  ------                                               --------------  -----   
 0   Unnamed: 0                                           10677 non-null  int64   
 1   STATE/UT                                             10677 non-null  object  
 2   DISTRICT                                             10677 non-null  category
 3   Year                                                 10677 non-null  int64   
 4   Rape                                                 10677 non-null  int64   
 5   Kidnapping and Abduction                             10677 non-null  int64   
 6   Dowry Deaths                                         10677 non-null  int64   
 7   Assault on women with intent to outrage her modesty  10677 non-null  int64   
 8   Insult to modesty of Women                           106

In [ ]:
DISTRICT категориальный, Assault on women with intent to outrage her modesty числовой количественный, Dowry Deaths числовой,Rape числовой,
 в моем датасете нет прям отдельных данных, которые можно было бы назвать порядковыми, поэтому Unnamed: 0	я возьму как числовой-порядковый

Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [ ]:
#Фильтр колонки 1
df[df["Year"] > 2008]

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
5869         5869  ANDHRA PRADESH           ADILABAD  2009    67   
5870         5870  ANDHRA PRADESH          ANANTAPUR  2009    22   
5871         5871  ANDHRA PRADESH           CHITTOOR  2009    33   
5872         5872  ANDHRA PRADESH           CUDDAPAH  2009    27   
5873         5873  ANDHRA PRADESH          CYBERABAD  2009    82   
...           ...             ...                ...   ...   ...   
10672         832     Lakshadweep        Lakshadweep  2014     1   
10673         833     Lakshadweep  Total District(s)  2014     1   
10674         834      Puducherry           Karaikal  2014     3   
10675         835      Puducherry         Puducherry  2014     7   
10676         836      Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
5869                         56            13   
5870                         60            14   
5871                         46            19   
5872                         47            17   
5873                        113            46   
...                         ...           ...   
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
5869                                                 192     
5870                                                 157     
5871                                                 120     
5872                                                 165     
5873                                                 173     
...                                                  ...     
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
5869                          163                                  428   
5870                           13                                  143   
5871                          268                                  333   
5872                           52                                  124   
5873                          155                                  991   
...                           ...                                  ...   
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
5869                      0  
5870                      0  
5871                      0  
5872                      0  
5873                      0  
...                     ...  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0  

[4808 rows x 11 columns]

In [ ]:
#Фильтр колонки 2
df[df["Insult to modesty of Women"] < 120]

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
0               0  ANDHRA PRADESH           ADILABAD  2001    50   
1               1  ANDHRA PRADESH          ANANTAPUR  2001    23   
2               2  ANDHRA PRADESH           CHITTOOR  2001    27   
3               3  ANDHRA PRADESH           CUDDAPAH  2001    20   
4               4  ANDHRA PRADESH      EAST GODAVARI  2001    23   
...           ...             ...                ...   ...   ...   
10672         832     Lakshadweep        Lakshadweep  2014     1   
10673         833     Lakshadweep  Total District(s)  2014     1   
10674         834      Puducherry           Karaikal  2014     3   
10675         835      Puducherry         Puducherry  2014     7   
10676         836      Puducherry  Total District(s)  2014    10   

       Kidnapping and Abduction  Dowry Deaths  \
0                            30            16   
1                            30             7   
2                            34            14   
3                            20            17   
4                            26            12   
...                         ...           ...   
10672                         0             0   
10673                         0             0   
10674                         1             0   
10675                         6             1   
10676                         7             1   

       Assault on women with intent to outrage her modesty  \
0                                                    149     
1                                                    118     
2                                                    112     
3                                                    126     
4                                                    109     
...                                                  ...     
10672                                                  1     
10673                                                  1     
10674                                                 12     
10675                                                 20     
10676                                                 32     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
0                              34                                  175   
1                              24                                  154   
2                              83                                  186   
3                              38                                   57   
4                              58                                  247   
...                           ...                                  ...   
10672                           2                                    0   
10673                           2                                    0   
10674                           1                                    1   
10675                           7                                    3   
10676                           8                                    4   

       Importation of Girls  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  
...                     ...  
10672                     0  
10673                     0  
10674                     0  
10675                     0  
10676                     0  

[10273 rows x 11 columns]

In [ ]:
#Фильтр колонки 3
df[df["Dowry Deaths"] > 30]

Unnamed: 0        STATE/UT           DISTRICT  Year  Rape  \
8               8  ANDHRA PRADESH         KARIMNAGAR  2001    56   
14             14  ANDHRA PRADESH           NALGONDA  2001    35   
18             18  ANDHRA PRADESH        RANGA REDDY  2001    72   
26             26  ANDHRA PRADESH           WARANGAL  2001    53   
28             28  ANDHRA PRADESH              TOTAL  2001   871   
...           ...             ...                ...   ...   ...   
10622         782     West Bengal            Hooghly  2014    45   
10631         791     West Bengal        Murshidabad  2014   252   
10632         792     West Bengal              Nadia  2014   106   
10640         800     West Bengal  Total District(s)  2014  1466   
10671         831        Delhi UT  Total District(s)  2014  2096   

       Kidnapping and Abduction  Dowry Deaths  \
8                            49            62   
14                           19            31   
18                           83            37   
26                           25            52   
28                          765           420   
...                         ...           ...   
10622                       213            34   
10631                       457            36   
10632                       336            50   
10640                      4976           501   
10671                      4034           153   

       Assault on women with intent to outrage her modesty  \
8                                                    414     
14                                                   188     
18                                                   113     
26                                                   241     
28                                                  3544     
...                                                  ...     
10622                                                222     
10631                                                523     
10632                                                181     
10640                                               5670     
10671                                               4322     

       Insult to modesty of Women  Cruelty by Husband or his Relatives  \
8                              81                                  224   
14                             59                                  214   
18                             55                                  421   
26                              0                                  165   
28                           2271                                 5791   
...                           ...                                  ...   
10622                           6                                 1338   
10631                          16                                 2439   
10632                           8                                 1827   
10640                         543                                23278   
10671                        1361                                 3173   

       Importation of Girls  
8                         0  
14                        0  
18                        7  
26                        0  
28                        7  
...                     ...  
10622                     0  
10631                     0  
10632                     1  
10640                     4  
10671                     0  

[1184 rows x 11 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [ ]:
#df.groupby(["Importation of Girls"]).mean()
df.groupby(['Importation of Girls']).mean(numeric_only=True)
#df["Importation of Girls"].mean()

Unnamed: 0         Year         Rape  \
Importation of Girls                                          
0                     3867.384172  2007.703722    48.174440   
1                     3737.386792  2007.037736   300.113208   
2                     4003.617021  2008.978723   319.042553   
3                     4681.708333  2007.375000   417.333333   
4                     4798.062500  2007.937500   241.250000   
5                     4380.000000  2008.611111   787.277778   
6                     4633.666667  2007.000000   845.166667   
7                      713.000000  2006.600000  1282.400000   
8                     4679.285714  2007.000000   576.571429   
9                     5014.428571  2010.428571   271.857143   
10                    7711.666667  2011.000000   353.000000   
11                    2857.000000  2004.000000   108.000000   
12                    5954.142857  2008.428571   744.142857   
14                     573.000000  2001.000000   423.000000   
15                    5221.000000  2008.000000    32.000000   
21                    1548.000000  2003.000000    25.000000   
22                    4855.500000  2007.500000   657.000000   
23                    5981.000000  2009.000000    18.000000   
25                    3011.500000  2005.000000    37.000000   
27                    4556.000000  2007.000000    57.500000   
29                     107.000000  2001.000000    19.000000   
31                    5991.000000  2009.000000   929.000000   
32                    8536.000000  2012.000000   621.000000   
34                    1690.500000  2003.000000    14.000000   
35                    2281.000000  2004.000000  1390.000000   
36                    1695.500000  2003.000000   797.000000   
37                    1553.000000  2003.000000   985.000000   
38                     831.000000  2002.000000  1040.000000   
39                    2740.000000  2004.500000   409.000000   
42                    3746.000000  2006.000000  1232.000000   
45                    7810.000000  2011.000000  3406.000000   
48                    3005.000000  2005.000000    28.000000   
56                    4486.000000  2007.000000  1555.000000   
60                    3590.000000  2005.000000   148.000000   
61                    3597.000000  2005.000000  1686.000000   
74                    3013.000000  2005.000000  1147.000000   
83                     115.000000  2001.000000   888.000000   

                      Kidnapping and Abduction  Dowry Deaths  \
Importation of Girls                                           
0                                    59.126166     16.369843   
1                                   305.216981     69.867925   
2                                   416.829787     71.638298   
3                                   401.750000    183.416667   
4                                   593.437500    141.812500   
5                                   815.222222    300.833333   
6                                   391.000000    237.833333   
7                                  1630.800000    489.000000   
8                                   873.571429    306.714286   
9                                   606.285714     76.285714   
10                                 1030.333333    478.333333   
11                                   80.000000     30.000000   
12                                  928.571429    247.428571   
14                                  607.000000    191.000000   
15                                   62.000000     60.000000   
21                                   18.500000     28.000000   
22                                  916.000000    620.000000   
23                                   11.000000     41.000000   
25                                   18.500000     36.500000   
27                                   36.500000     29.500000   
29                                   29.000000     11.000000   
31                                 1986.000000   1295.000000   
32                           

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация